matsim/src/main/java/org/matsim/core/population/io/ParallelPopulationReaderMatsimV6.java

```diff
@@ -0,0 +1,222 @@
+			log.warn("Population streaming is activated - cannot use " + ParallelPopulationReaderMatsimV6.class.getName() + "!");
+
+			this.isPopulationStreaming = true;
+			this.numThreads = 1;
+			this.queue = null;
+
+		} else {
+			isPopulationStreaming = false;
+
+			if (scenario.getConfig().global().getNumberOfThreads() > 0) {
+				this.numThreads = scenario.getConfig().global().getNumberOfThreads();
+			} else this.numThreads = 1;
+
+			this.queue = new LinkedBlockingQueue<>();
+		}
+	}
+
+	private void initThreads() {
+		threads = new Thread[numThreads];
+		for (int i = 0; i < numThreads; i++) {
+
+			ParallelPopulationReaderMatsimV6Runner runner =
+					new ParallelPopulationReaderMatsimV6Runner(
+							this.inputCRS,
+							this.targetCRS,
+							this.scenario,
+							this.queue);
+
+			Thread thread = new Thread(runner);
+			thread.setDaemon(true);
+			thread.setName(ParallelPopulationReaderMatsimV6Runner.class.toString() + i);
+			threads[i] = thread;
+			thread.start();
+		}
+	}
+
+	@Override
+	public void startTag(String name, Attributes atts, Stack<String> context) {
+		//Reached first time a person
+		if (PERSON.equals(name) && !this.reachedPersons) {
+			this.reachedPersons = true;
+
+			if (!isPopulationStreaming && this.threads == null) {
+				log.info("Start parallel population reading...");
+				initThreads();
+			}
+		}
+
+		// if population streaming is activated, use non-parallel reader
+		if (isPopulationStreaming || !this.reachedPersons) {
+			super.startTag(name, atts, context);
+			return;
+		}
+
+
+		// If it is a new person, create a new person and a list for its attributes.
+		if (PERSON.equals(name)) {
+			Person person = this.plans.getFactory().createPerson(Id.create(atts.getValue(ATTR_PERSON_ID), Person.class));
+			currentPersonXmlData = new ArrayList<>();
+			PersonTag personTag = new PersonTag();
+			personTag.person = person;
+			currentPersonXmlData.add(personTag);
+			this.plans.addPerson(person);
+		} else {
+
+			// Create a new start tag and add it to the person data.
+			Stack<String> contextCopy = new Stack<>();
+			contextCopy.addAll(context);
+			StartTag tag = new StartTag();
+			tag.name = name;
+			tag.context = contextCopy;
+			tag.atts = new AttributesImpl(atts);    // We have to create copies of the attributes because the object is re-used by the parser!
+			currentPersonXmlData.add(tag);
+		}
+	}
+
+	@Override
+	public void endTag(String name, String content, Stack<String> context) {
+
+		// if population streaming is activated, use non-parallel reader
+		// or if not reached the persons in the xml
+		if (isPopulationStreaming || !this.reachedPersons) {
+			super.endTag(name, content, context);
+			return;
+		}
+
+		// End of population reached
+		if (POPULATION.equals(name)) {
+			// signal the threads that they should end parsing
+			for (int i = 0; i < this.numThreads; i++) {
+				List<Tag> list = new ArrayList<>();
+				list.add(new EndProcessingTag());
+				this.queue.add(list);
```

Comment:
modern Java allows for a more efficient (effect will be miniscule) and cleaner code:
```java
this.queue.add(List.of(new EndProcessingTag()));
```

Reply:
Done